lib/uksched/thread.c

```diff
@@ -41,27 +41,21 @@
 #include <uk/assert.h>
 #include <uk/arch/tls.h>
 
-/* Pushes the specified value onto the stack of the specified thread */
-static void stack_push(unsigned long *sp, unsigned long value)
-{
-	*sp -= sizeof(unsigned long);
-	*((unsigned long *) *sp) = value;
-}
+typedef void (*capsule_fn)(void *);
 
-static void init_sp(unsigned long *sp, char *stack,
-		void (*function)(void *), void *data)
+static __noreturn void thread_capsule(long funcp, long argp)
```

Comment:
Now here is what I found, this is how a thread is created for `pthread-embedded`
```
struct pte_entry_capsule {
        pte_osThreadEntryPoint entry_point;
        void *argv;
};
...
pte_osThreadCreate() { 
  ...
  capsule.entry_point = entry_point;
  capsule.argv        = argv;
  ...
  th = uk_thread_create_attr(NULL, NULL, PTE_CAPSULE_MAGIC, &capsule);
  ...
}
```

When the execution reaches this function, the function pointer will point to `PTE_CAPSULE_MAGIC`, which is just garbage in this case.

Reply:
You are right, in fact, we currently decided to break newlib and pthread embedded. The idea is to come back to it later, this is why we have currently `stagin-muslpreview` as branch for these PRs. newlib and especially `pthread-embedded` need an updated integration. However, this would be much cleaner with the latest APIs.
Additionally, this capsule is a temporary solution for implementing libcontext with the current `uksched` API. The upcoming scheduler PR with the new API will properly integrate `libcontext` and will make this capsule unnecessary.